### PR TITLE
Plugins: Add success and error notices for uploads

### PR DESCRIPTION
**Pull request by seear**
screen shot 2017-07-21 at 11 43 25
screen shot 2017-07-21 at 11 42 57

**To Test**
* Checkout this branch or use calypso.live
* Go to /plugins/upload/{jetpack_site}
* Upload a plugin
* Try uploading the same plugin again
* Try dragging in a file that is not a plugin zip

**Expected**
* An appropriate notice should be displayed

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 1]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+		{ duration: 5000 }
+	) );
+};
+
+const showErrorNotice = ( dispatch, error ) => {
+	const knownErrors = {
+		exists: translate( 'Upload problem: Plugin already installed on site.' ),
+		'too large': translate( 'Upload problem: Plugin zip must be smaller than 10MB.' ),
+		incompatible: translate( 'Upload problem: Incompatible plugin.' ),
+		unsupported_mime_type: translate( 'Upload problem: Not a valid zip file.' ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Upload problem: Not a valid zip file' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/plugin-upload-notices)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 2]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+		} ),
+		{ duration: 5000 }
+	) );
+};
+
+const showErrorNotice = ( dispatch, error ) => {
+	const knownErrors = {
+		exists: translate( 'Upload problem: Plugin already installed on site.' ),
+		'too large': translate( 'Upload problem: Plugin zip must be smaller than 10MB.' ),
+		incompatible: translate( 'Upload problem: Incompatible plugin.' ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Upload problem: Incompatible theme.' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 3]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+		translate( 'Successfully uploaded plugin %(name)s.', {
+			args: { name }
+		} ),
+		{ duration: 5000 }
+	) );
+};
+
+const showErrorNotice = ( dispatch, error ) => {
+	const knownErrors = {
+		exists: translate( 'Upload problem: Plugin already installed on site.' ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Upload problem: Theme already installed on site.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/plugins/new/index.js (hunk 1 above)
Comment (seear) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: 👍

File: client/state/data-layer/wpcom/sites/plugins/new/index.js (hunk 2 above)
Comment (seear) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: 👎

File: client/state/data-layer/wpcom/sites/plugins/new/index.js (hunk 3 above)
Comment (seear) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: 👎

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 4]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
 	}, action ) );
 };
 
+const showSuccessNotice = ( dispatch, { name } ) => {
+	dispatch( successNotice(
+		translate( 'Successfully uploaded plugin %(name)s.', {
```
Comment (tyxla) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: I can see there's a dot at the end of the success notice in the code, but it's not in the screenshot. Perhaps you put the dot after you made the screenshot? :)

Comment (seear) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: correct :)

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 5]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+
+const showErrorNotice = ( dispatch, error ) => {
+	const knownErrors = {
+		exists: translate( 'Upload problem: Plugin already installed on site.' ),
+		'too large': translate( 'Upload problem: Plugin zip must be smaller than 10MB.' ),
+		incompatible: translate( 'Upload problem: Incompatible plugin.' ),
+		unsupported_mime_type: translate( 'Upload problem: Not a valid zip file.' ),
+	};
+	const errorString = toLower( error.error + error.message );
+	const knownError = find( knownErrors, ( v, key ) => includes( errorString, key ) );
```
Comment (seear) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: The errors come from various different places—some from the jetpack plugin, others from wpcom—so there is not a consistency of error codes. Smushing the code and message together and doing a fairly fuzzy match allows us to match all the known errors with one lookup. 

Also, with themes we found we needed to tweak things as new error reports came in.

**Comment by seear** [comment]
> Do we need a copy review?

@michelleweber want to take a quick look at the new strings here?

File: client/state/data-layer/wpcom/sites/plugins/new/index.js (hunk 5 above)
Comment (tyxla) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Sounds like a great approach in that case :+1:

**Comment by michelleweber** [comment]
I'd love it if we could make these sound a little more like natural speech:

- The X plugin has been successfully uploaded
or
- You've successfully uploaded the X plugin. (My preference)

and

- This plugin is already installed on your site.

(I don't think you need the "upload problem" -- the red color + exclamation mark + message communicates that pretty clearly.)

**Comment by keoshi** [comment]
I love all of that, and my favorite is also the second option. Thanks @michelleweber !

**Comment by seear** [comment]
Thanks @michelleweber! Yes, that is much better, I've updated the strings appropriately.

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 6]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+		dispatch( errorNotice( knownError ) );
+		return;
+	}
+	if ( error.error ) {
+		dispatch( errorNotice( translate( 'Upload problem: %(error)s.', {
+			args: { error: error.error }
+		} ) ) );
+		return;
+	}
+	dispatch( errorNotice( translate( 'Problem installing the plugin.' ) ) );
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Hi! I've found a possible matching string that has already been translated **38** times:
`translate( 'Installing the plugin.' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/state/data-layer/wpcom/sites/plugins/new/index.js [hunk 7]
```diff
@@ -21,6 +28,38 @@ export const uploadPlugin = ( { dispatch }, action ) => {
...
+		translate( "You've successfully uploaded the %(name)s plugin.", {
+			args: { name }
+		} ),
+		{ duration: 5000 }
+	) );
+};
+
+const showErrorNotice = ( dispatch, error ) => {
+	const knownErrors = {
+		exists: translate( 'This plugin is already installed on your site.' ),
```
Comment (a8ci18n) [inline client/state/data-layer/wpcom/sites/plugins/new/index.js]: Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Plugin is already installed.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Pull request merged**